add tvu port binding info to docs (#4948)
anza-xyz/agave: docs/src/validator/tvu.md

---

TVU (Transaction Validation Unit) is the logic of the validator
responsible for propagating blocks between validators and ensuring that
those blocks' transactions reach the replay stage. Its principal external

> interface is the turbine protocol.

Reviewer: not sure i completely agree with this. TVU validates the transactions it has received from turbine. turbine's role is propagating blocks between valdiators. TVU's role is validating those blocks

Author: Well in this case the picture is all wrong (and it predates this PR)

Reviewer: TVU is also responnsible for validator the blocks.

Reviewer: possible reword here:
```suggestion
TVU (Transaction Validation Unit) is the validator component responsible for receiving, verifying, and propagating blocks between validators. It ensures transactions from blocks reach the replay stage for execution. Its principal external interface for block propagation is the Turbine protocol, which distributes blocks efficiently across the network.
```

![TVU Block Diagram](/img/tvu.svg)

## Retransmit Stage

![Retransmit Block Diagram](/img/retransmit_stage.svg)

## TVU sockets

Externally, TVU UDP receiver appears to bind to one port, typically 8002 UDP.
Internally, TVU is actually bound with multiple sockets to improve kernel's handling of the packet queues.

> **NOTE:** TPU sockets use similar logic

A node advertises one external ip/port for TVU while binding multiple sockets to that same port using SO_REUSEPORT:

```rust

> let (tvu_port, tvu_sockets) = multi_bind_in_range_with_config(
>             bind_ip_addr,
>             port_range,
>             socket_config_reuseport,
>             num_tvu_sockets.get(),
>         )
>         .expect("tvu multi_bind");

Reviewer: ```suggestion
let (tvu_port, tvu_sockets) = multi_bind_in_range_with_config(
    bind_ip_addr,
    port_range,
    socket_config_reuseport,
    num_tvu_sockets.get(),
)
.expect("tvu multi_bind");
```

```

 `multi_bind_in_range_with_config` sets `SO_REUSEPORT`. This means that other nodes only need to know about the one ip/port pair for TVU (similar principle applies in the case of TPU UDP sockets). The kernel distributes the incoming packets to all sockets bound to that port, and each socket can be serviced by a different thread.

> **NOTE:** TVU QUIC socket does not use `SO_REUSEPORT`, but otherwise works similarly to UDP

> The TVU socket information is published via Gossip and is available in `ContactInfo` struct.

Reviewer: ```suggestion
The TVU socket information is published via Gossip and is available in the `ContactInfo` struct.
```

## Retransmit Stage

![Retransmit Block Diagram](/img/retransmit_stage.svg)

> 

Reviewer: remove extra line

## TVU sockets

> Externally, TVU appears to bind to one port, typically 8002 UDP. Internally, TVU is actually bound with multiple sockets

Reviewer: so we have both UDP and QUIC ports for TVU. Probably makes sense to differentiate these two somewhere. UDP multibind with reuseport works on OS level. For quic tvu, we only bind to one socket. and the multiplexing is done at the protocol level.

Author: done

> to improve kernel's handling of the packet queues. It is setup so that a node can advertise one external ip/port for TVU/TPU. We're binding multiple sockets to the same port using SO_REUSEPORT:

Reviewer: > external ip/port for TVU/TPU.

would maybe leave out TPU inline here. Maybe we can add something right after or even at the top saying TPU runs on similar port management logic

Author: added note about TPU

```rust
let (tvu_port, tvu_sockets) = multi_bind_in_range_with_config(
            bind_ip_addr,
            port_range,
            socket_config_reuseport,
            num_tvu_sockets.get(),
        )
        .expect("tvu multi_bind");
```

 `multi_bind_in_range_with_config` sets `SO_REUSEPORT`. This means that other nodes only need to know about the one ip/port pair for TVU (similar principle applies in the case of TPU UDP sockets). The kernel distributes the incoming packets to all sockets bound to that port, and each socket can be serviced by a different thread.

> 

Reviewer: remove extra line

The TVU socket information is published via Gossip and is available in `ContactInfo` struct.

> To set a tvu socket, the node can call set_tvu() which is created by the macro `set_socket!`. For example:

Reviewer: ```suggestion
To set a TVU socket, the node calls `set_tvu(...)`. The `set_tvu()` method is created by the macro `set_socket!`. For example:
```

```rust
info.set_tvu(QUIC, (addr, tvu_quic_port)).unwrap();
```

> `set_tvu()` will call `set_socket()`
> 
> under the hood in the actual `ContactInfo` CRDS all sockets are identified by a tag/key like these:
> 
> ```rust
> const SOCKET_TAG_TVU: u8 = 10;       // For UDP
> const SOCKET_TAG_TVU_QUIC: u8 = 11;  // For QUIC
> ```

Reviewer: ```suggestion
Under the hood, `set_tvu()` calls `set_socket()`.

In the `ContactInfo` struct, all sockets are identified by a tag/key, e.g.:

```rust
const SOCKET_TAG_TVU: u8 = 10;       // For UDP
const SOCKET_TAG_TVU_QUIC: u8 = 11;  // For QUIC
```
```

>  * `set_socket()` will create a `SocketEntry` and store that into `ContactInfo::sockets`
>  * `set_socket()` will also update the `ContactInfo::cache`

Reviewer: ```suggestion
 * `set_socket()` creates a `SocketEntry` and stores that into `ContactInfo::sockets`
 * `set_socket()` updates `ContactInfo::cache`
```

```rust
cache: [SocketAddr; SOCKET_CACHE_SIZE]
```

> the cache is purely for quick lookups and optimization. it is not serialized and sent to peer nodes.

Reviewer: ```suggestion
`cache` is purely for quick lookups and optimization; it is not serialized and sent to peer nodes.
```

> But, `SocketEntry` will be serialized and sent to peer nodes in the `ContactInfo` CRDS. On the receiving end, the `get_socket!` macro will return the TVU port.

Reviewer: ```suggestion
`SocketEntry` is serialized and sent to peer nodes within the gossip message type `CrdsData::ContactInfo`. Upon receiving the `ContactInfo`, the peer node calls the `get_socket!` macro to retrieve the TVU port associated with the node..
```

> so you can call:
> ```rust
> get_socket!(tvu, SOCKET_TAG_TVU, SOCKET_TAG_TVU_QUIC);
> ```
> to retrieve the TVU ports of the remote node.

Reviewer: ```suggestion
For example, to retrieve the TVU ports of the remote node, the peer node calls:
```rust
get_socket!(tvu, SOCKET_TAG_TVU, SOCKET_TAG_TVU_QUIC);
```
```

## Retransmit Stage

![Retransmit Block Diagram](/img/retransmit_stage.svg)

## TVU sockets

Externally, TVU UDP receiver appears to bind to one port, typically 8002 UDP.
Internally, TVU is actually bound with multiple sockets

> to improve kernel's handling of the packet queues.

Author: Cargo fmt does not touch markdown files afaik. I formatted it like this to match the layout of the text above. When rendering for online presentation it does reformatting anyway.

> **NOTE:** TPU sockets use similar logic

> It is setup so that a node can advertise one external ip/port for TVU.
>  We're binding multiple sockets to the same port using SO_REUSEPORT:

Reviewer: ```suggestion
A node advertises one external ip/port for TVU while binding multiple sockets to that same port using SO_REUSEPORT:
```

```rust
let (tvu_port, tvu_sockets) = multi_bind_in_range_with_config(
            bind_ip_addr,
            port_range,
            socket_config_reuseport,
            num_tvu_sockets.get(),
        )
        .expect("tvu multi_bind");
```

 `multi_bind_in_range_with_config` sets `SO_REUSEPORT`. This means that other nodes only need to know about the one ip/port pair for TVU (similar principle applies in the case of TPU UDP sockets). The kernel distributes the incoming packets to all sockets bound to that port, and each socket can be serviced by a different thread.

> **NOTE:** TVU QUIC socket does not use `SO_REUSEPORT`, but otherwise works similarly to UDP

The TVU socket information is published via Gossip and is available in `ContactInfo` struct.
To set a TVU socket, the node calls `set_tvu(...)`. The `set_tvu()` method is created by the macro `set_socket!`. For example:

```rust
info.set_tvu(QUIC, (addr, tvu_quic_port)).unwrap();
info.set_tvu(UDP, (addr, tvu_udp_port)).unwrap();
```

Under the hood, `set_tvu()` calls `set_socket()`.

In the `ContactInfo` struct, all sockets are identified by a tag/key, e.g.:

> ```rust
> const SOCKET_TAG_TVU: u8 = 10;       // For UDP
> const SOCKET_TAG_TVU_QUIC: u8 = 11;  // For QUIC
> 

Reviewer: ```suggestion
```rust
const SOCKET_TAG_TVU: u8 = 10;       // For UDP
const SOCKET_TAG_TVU_QUIC: u8 = 11;  // For QUIC
```
```
^ note the missing code block closing "`s"

Reviewer: ok this formatted weird. but basically, we're missing a ``` at the end of this block, so formatting is weird

 * `set_socket()` creates a `SocketEntry` and stores that into `ContactInfo::sockets`
 * `set_socket()` updates `ContactInfo::cache`

```rust
cache: [SocketAddr; SOCKET_CACHE_SIZE]
```

`cache` is purely for quick lookups and optimization; it is not serialized and sent to peer nodes.

> `SocketEntry` is serialized and sent to peer nodes within the gossip message type `CrdsData::ContactInfo`. Upon receiving the `ContactInfo`, the peer node calls the `get_socket!` macro to retrieve the TVU port associated with the node..

Reviewer: ```suggestion
`SocketEntry` is serialized and sent to peer nodes within the gossip message type `CrdsData::ContactInfo`. Upon receiving the `ContactInfo`, the peer node calls the `get_socket!` macro to retrieve the TVU port associated with the node.
```
